relayer/relays/beefy/on-demand-sync.go

```diff
@@ -72,101 +83,93 @@ func (relay *OnDemandRelay) Start(ctx context.Context) error {
+	relay.gatewayContractV1 = gatewayContractV1
 
 	relay.tokenBucket.Start(ctx)
 
 	for {
-		sleep(ctx, time.Minute*1)
-		log.Info("Starting check")
-
-		paraNonce, ethNonce, err := relay.queryNonces(ctx)
-		if err != nil {
-			if errors.Is(err, context.Canceled) {
-				return nil
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(time.Second * 30):
+			log.Info("Starting check")
+			err = relay.syncV1(ctx)
+			if err != nil {
+				return fmt.Errorf("sync for v1 message: %w", err)
+			}
+			err = relay.syncV2(ctx)
+			if err != nil {
+				return fmt.Errorf("sync for v2 message: %w", err)
 			}
-			log.WithError(err).Error("Query nonces")
-			continue
 		}
+	}
+}
 
-		log.WithFields(log.Fields{
-			"paraNonce": paraNonce,
-			"ethNonce":  ethNonce,
-		}).Info("Nonces checked")
+func (relay *OnDemandRelay) syncV1(ctx context.Context) error {
+	paraNonce, ethNonce, err := relay.queryNonces(ctx)
+	if err != nil {
+		return fmt.Errorf("Query nonces: %w", err)
+	}
 
-		if paraNonce > ethNonce {
+	log.WithFields(log.Fields{
+		"paraNonce": paraNonce,
+		"ethNonce":  ethNonce,
+	}).Info("Nonces checked")
 
-			// Check if we are rate-limited
-			if !relay.tokenBucket.TryConsume(1) {
-				log.Info("Rate-limit exceeded")
-				continue
-			}
+	if paraNonce > ethNonce {
 
-			log.Info("Performing sync")
+		// Check if we are rate-limited
+		if !relay.tokenBucket.TryConsume(1) {
+			return fmt.Errorf("Rate-limit exceeded")
+		}
 
-			beefyBlockHash, err := relay.relaychainConn.API().RPC.Beefy.GetFinalizedHead()
-			if err != nil {
-				if errors.Is(err, context.Canceled) {
-					return nil
-				}
-				log.WithError(err).Error("Fetch latest beefy block hash")
-				continue
-			}
+		log.Info("Performing sync v1")
 
-			header, err := relay.relaychainConn.API().RPC.Chain.GetHeader(beefyBlockHash)
-			if err != nil {
-				if errors.Is(err, context.Canceled) {
-					return nil
-				}
-				log.WithError(err).Error("Fetch latest beefy block header")
-				continue
-			}
-
-			err = relay.sync(ctx, uint64(header.Number))
-			if err != nil {
-				if errors.Is(err, context.Canceled) {
-					return nil
-				}
-				log.WithError(err).Error("Sync failed")
-				continue
-			}
+		beefyBlockHash, err := relay.relaychainConn.API().RPC.Beefy.GetFinalizedHead()
+		if err != nil {
+			return fmt.Errorf("Fetch latest beefy block hash: %w", err)
+		}
 
-			log.Info("Sync completed")
+		header, err := relay.relaychainConn.API().RPC.Chain.GetHeader(beefyBlockHash)
+		if err != nil {
+			return fmt.Errorf("Fetch latest beefy block header: %w", err)
+		}
 
-			relay.waitUntilMessagesSynced(ctx, paraNonce)
+		err = relay.sync(ctx, uint64(header.Number))
+		if err != nil {
+			return fmt.Errorf("Sync failed: %w", err)
 		}
+
+		log.Info("Sync completed")
+
+		relay.waitUntilMessagesSynced(ctx, paraNonce)
 	}
+	return nil
 }
 
 func (relay *OnDemandRelay) waitUntilMessagesSynced(ctx context.Context, paraNonce uint64) {
-	sleep(ctx, time.Minute*10)
+	waitingTime := 0
```

Comment:
This new approach where the waiting times out creates a problem: In the next iteration of the loop within `relay.Start`, the same commitment will be synced, costing money. And this will continue forever until the message relayer finally syncs the user message.

So for V1, we should rather have `waitUntilMessagesSynced` wait forever (i.e revert these changes). Since we run the user message relayers we need to keep them up an running always.

Reply:
https://github.com/Snowfork/snowbridge/pull/1449/commits/e09f957af4bd3077785c348d0a82aae1530897f1